File: go/filefs/copy.go
package filefs

// Copy duplicates a file (C: FileFS_copy).
// return: 0:ok,1-err,2-from name format err,3-to path format err,4-from name not exist,5-to file exist
func (ffs *FileFS) Copy(fromFilename, toFilename string) int {
	if ffs == nil || ffs.Fp == nil || fromFilename == "" || toFilename == "" {
		return 1
	}

	lenN := len(fromFilename)
	if fromFilename[lenN-1] == '/' {
		return 2
	}
	var blockindex uint32
	var start int
	if fromFilename[0] == '/' {
		blockindex = 1
		start = 1
	} else {
		if ffs.Tmp.State == 0 {
			blockindex = ffs.PwdBlockindex
		} else {
			blockindex = ffs.Tmp.PwdBlockindex
		}
		start = 0
	}
	var s [BlockNameMaxSize + 2]byte
	slen := 0
	for i := start; i < lenN; i++ {
		if fromFilename[i] == '/' {
			if slen == 0 {
				continue
			}
			if i == lenN-1 {
				break
			}
			index := ffs.findPathBlockindex(blockindex, string(s[:slen]))
			if index < 1 {
				return 2
			}
			blockindex = index
			slen = 0
			continue
		}
		s[slen] = fromFilename[i]
		slen++
		if slen > BlockNameMaxSize {
			return 2
		}
	}
	if slen > BlockNameMaxSize {
		return 2
	}
	fromLastname := string(s[:slen])
	if fromLastname == "." || fromLastname == ".." {
		return 2
	}
	fromBlockindex := blockindex

	lenN = len(toFilename)
	if toFilename[lenN-1] == '/' {
		return 3
	}
	if toFilename[0] == '/' {
		blockindex = 1
		start = 1
	} else {
		if ffs.Tmp.State == 0 {
			blockindex = ffs.PwdBlockindex
		} else {
			blockindex = ffs.Tmp.PwdBlockindex
		}
		start = 0
	}
	slen = 0
	for i := start; i < lenN; i++ {
		if toFilename[i] == '/' {
			if slen == 0 {
				continue
			}
			if i == lenN-1 {
				break
			}
			index := ffs.findPathBlockindex(blockindex, string(s[:slen]))
			if index < 1 {
				return 3
			}
			blockindex = index
			slen = 0
			continue
		}
		s[slen] = toFilename[i]
		slen++
		if slen > BlockNameMaxSize {
			return 3
		}
	}
	if slen > BlockNameMaxSize {
		return 3
	}
	toLastname := string(s[:slen])
	if toLastname == "." || toLastname == ".." {
		return 3
	}
	toBlockindex := blockindex

	var b4 [4]byte
	var b2 [2]byte
	var fromBlock [BlockSize]byte
	if !ffs.readblock(fromBlockindex, fromBlock[:]) {
		return 1
	}
	copy(b4[:], fromBlock[12+1+14+4:12+1+14+8])
	fromStopBlockindex := B4toU32(b4[:])
	copy(b2[:], fromBlock[12+1+14+4+4:12+1+14+4+6])
	fromOffset := B2toU16(b2[:])

	var fromFileStart, fromFileStop uint32
	var fromFileOffset uint16
	flag := false
	index := fromBlockindex
	for {
		k := BlockHead
		for i := 0; i < BlockItemMaxCount; i++ {
			state := fromBlock[k]
			k++
			copy(s[:BlockNameMaxSize], fromBlock[k:k+BlockNameMaxSize])
			k += BlockNameMaxSize
			if cstrFromFixed(s[:BlockNameMaxSize+1]) != fromLastname {
				k += 10
				if index == fromStopBlockindex && k+1 >= int(fromOffset) {
					return 4
				}
				continue
			}
			dirFile := state & 0x01
			if dirFile != 1 {
				return 2
			}
			copy(b4[:], fromBlock[k:k+4])
			k += 4
			fromFileStart = B4toU32(b4[:])
			copy(b4[:], fromBlock[k:k+4])
			k += 4
			fromFileStop = B4toU32(b4[:])
			copy(b2[:], fromBlock[k:k+2])
			fromFileOffset = B2toU16(b2[:])
			flag = true
			break
		}
		if flag {
			break
		}
		copy(b4[:], fromBlock[4:8])
		index = B4toU32(b4[:])
		if index == 0 {
			return 1
		}
		if !ffs.readblock(index, fromBlock[:]) {
			return 1
		}
	}

	var toBa [2]BlockArray
	toBaUsed := 0
	var toBlockHead, toBlockLast []byte
	var toBlockHeadIndex, toBlockLastIndex uint32

	var toBlock [BlockSize]byte
	if !ffs.readblock(toBlockindex, toBlock[:]) {
		return 1
	}
	copy(toBa[0].Block[:], toBlock[:])
	toBa[0].Blockindex = toBlockindex
	toBa[0].Active = 1
	toBlockHead = toBa[0].Block[:]
	toBlockHeadIndex = toBlockindex
	toBaUsed = 1

	copy(b4[:], toBlockHead[12+1+14+4:12+1+14+8])
	toStopBlockindex := B4toU32(b4[:])
	copy(b2[:], toBlockHead[12+1+14+4+4:12+1+14+4+6])
	toOffset := B2toU16(b2[:])

	if toStopBlockindex == toBlockHeadIndex {
		toBlockLast = toBlockHead
		toBlockLastIndex = toBlockHeadIndex
	} else {
		if !ffs.readblock(toStopBlockindex, toBa[1].Block[:]) {
			return 1
		}
		toBa[1].Blockindex = toStopBlockindex
		toBa[1].Active = 1
		toBlockLast = toBa[1].Block[:]
		toBlockLastIndex = toStopBlockindex
		toBaUsed++
	}

	flag = false
	index = toBlockHeadIndex
	for {
		k := BlockHead
		for i := 0; i < BlockItemMaxCount; i++ {
			k++
			copy(s[:BlockNameMaxSize], toBlock[k:k+BlockNameMaxSize])
			k += BlockNameMaxSize
			if cstrFromFixed(s[:BlockNameMaxSize+1]) != toLastname {
				k += 10
				if index == toStopBlockindex && k+1 >= int(toOffset) {
					flag = true
					break
				}
				continue
			}
			return 5
		}
		if flag {
			break
		}
		copy(b4[:], toBlock[4:8])
		index = B4toU32(b4[:])
		if index == 0 {
			return 1
		}
		if !ffs.readblock(index, toBlock[:]) {
			return 1
		}
	}

	if ffs.Tmp.State == 0 {
		ffs.tmpstart(1)
	}

	var block2 [BlockSize]byte
	var blockindex2 uint32
	var newToOffset uint16

	if toOffset < BlockSize {
		for i := 0; i < 25; i++ {
			toBlockLast[int(toOffset)+i] = 0
		}
		toBlockLast[toOffset] = 1
		copy(toBlockLast[toOffset+1:int(toOffset)+1+len(toLastname)], toLastname)
		newToOffset = toOffset + 25
		U16toB2(newToOffset, b2[:])
		copy(toBlockHead[BlockOffset:], b2[:])
	} else {
		blockindex2 = ffs.genblockindex()
		if blockindex2 == 0 {
			if ffs.Tmp.State == 1 {
				ffs.tmpstop()
			}
			return 1
		}
		for i := range block2 {
			block2[i] = 0
		}
		U32toB4(toBlockLastIndex, b4[:])
		copy(block2[8:], b4[:])
		block2[BlockHead] = 1
		copy(block2[BlockHead+1:BlockHead+1+len(toLastname)], toLastname)
		U32toB4(blockindex2, b4[:])
		copy(toBlockLast[4:], b4[:])
		newToOffset = BlockHead + 25
		U16toB2(newToOffset, b2[:])
		copy(toBlockHead[BlockOffset:], b2[:])
		copy(toBlockHead[BlockStopBlockindex:], b4[:])
	}

	var toFileStart, toFileStop uint32
	var toFileOffset uint16

	if fromFileStart > 0 {
		toFileOffset = fromFileOffset
		fromIndex := fromFileStart
		if !ffs.readblock(fromIndex, fromBlock[:]) {
			if ffs.Tmp.State == 1 {
				ffs.tmpstop()
			}
			return 1
		}
		copy(b4[:], fromBlock[4:8])
		fromNextIndex := B4toU32(b4[:])

		newBlockindex := ffs.genblockindex()
		if newBlockindex == 0 {
			if ffs.Tmp.State == 1 {
				ffs.tmpstop()
			}
			return 1
		}
		toFileStart = newBlockindex
		toFileStop = newBlockindex

		prevIndex := uint32(0)
		var newBlock [BlockSize]byte
		for {
			copy(newBlock[:], fromBlock[:])
			U32toB4(prevIndex, b4[:])
			copy(newBlock[8:], b4[:])

			if fromIndex == fromFileStop {
				toFileStop = newBlockindex
				if !ffs.writeblock(newBlockindex, newBlock[:]) {
					if ffs.Tmp.State == 1 {
						ffs.tmpstop()
					}
					return 1
				}
				break
			}

			prevIndex = newBlockindex
			newBlockindex = ffs.genblockindex()
			U32toB4(newBlockindex, b4[:])
			copy(newBlock[4:], b4[:])
			if !ffs.writeblock(prevIndex, newBlock[:]) {
				if ffs.Tmp.State == 1 {
					ffs.tmpstop()
				}
				return 1
			}

			fromIndex = fromNextIndex
			if !ffs.readblock(fromIndex, fromBlock[:]) {
				if ffs.Tmp.State == 1 {
					ffs.tmpstop()
				}
				return 1
			}
			copy(b4[:], fromBlock[4:8])
			fromNextIndex = B4toU32(b4[:])
		}
	}

	if toOffset < BlockSize {
		U32toB4(toFileStart, b4[:])
		copy(toBlockLast[toOffset+25-10:], b4[:])
		U32toB4(toFileStop, b4[:])
		copy(toBlockLast[toOffset+25-6:], b4[:])
		U16toB2(toFileOffset, b2[:])
		copy(toBlockLast[toOffset+25-2:], b2[:])
	} else {
		U32toB4(toFileStart, b4[:])
		copy(block2[BlockHead+25-10:], b4[:])
		U32toB4(toFileStop, b4[:])
		copy(block2[BlockHead+25-6:], b4[:])
		U16toB2(toFileOffset, b2[:])
		copy(block2[BlockHead+25-2:], b2[:])
		if !ffs.writeblock(blockindex2, block2[:]) {
			if ffs.Tmp.State == 1 {
				ffs.tmpstop()
			}
			return 1
		}
	}

	for i := 0; i < toBaUsed; i++ {
		if toBa[i].Active == 0 {
			continue
		}
		if !ffs.writeblock(toBa[i].Blockindex, toBa[i].Block[:]) {
			if ffs.Tmp.State == 1 {
				ffs.tmpstop()
			}
			return 1
		}
	}
	if ffs.Tmp.State == 1 {
		if !ffs.Commit() {
			return 1
		}
	}
	return 0
}

File: zig/src/copy.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");
const txn = @import("txn.zig");
const block_mod = @import("block.zig");

pub fn copy(self: anytype, from_filename: []const u8, to_filename: []const u8) i32 {
    if (self.fp == null or from_filename.len == 0 or to_filename.len == 0) return 1;
    if (from_filename[from_filename.len - 1] == '/') return 2;

    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (from_filename[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }
    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < from_filename.len) : (i += 1) {
        if (from_filename[i] == '/') {
            if (slen == 0) continue;
            if (i == from_filename.len - 1) break;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) return 2;
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = from_filename[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return 2;
    }
    var from_lastname_buf: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);
    @memcpy(from_lastname_buf[0..slen], s[0..slen]);
    const from_lastname = from_lastname_buf[0..slen];
    if (std.mem.eql(u8, from_lastname, ".") or std.mem.eql(u8, from_lastname, "..")) return 2;
    const from_blockindex = blockindex;

    if (to_filename[to_filename.len - 1] == '/') return 3;
    if (to_filename[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }
    slen = 0;
    i = start;
    while (i < to_filename.len) : (i += 1) {
        if (to_filename[i] == '/') {
            if (slen == 0) continue;
            if (i == to_filename.len - 1) break;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) return 3;
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = to_filename[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return 3;
    }
    var to_lastname_buf: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);
    @memcpy(to_lastname_buf[0..slen], s[0..slen]);
    const to_lastname = to_lastname_buf[0..slen];
    if (std.mem.eql(u8, to_lastname, ".") or std.mem.eql(u8, to_lastname, "..")) return 3;
    const to_blockindex = blockindex;

    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    var from_block = util.blockZero();
    if (!block_mod.readblock(self, from_blockindex, from_block[0..])) return 1;
    @memcpy(b4[0..], from_block[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const from_stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], from_block[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    const from_offset = util.b2ToU16(b2[0..]);

    var from_file_start: u32 = 0;
    var from_file_stop: u32 = 0;
    var from_file_offset: u16 = 0;
    var flag = false;
    var index = from_blockindex;
    while (true) {
        var k: usize = types.BlockHead;
        var item: usize = 0;
        while (item < types.BlockItemMaxCount) : (item += 1) {
            const state = from_block[k];
            k += 1;
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], from_block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..types.BlockNameMaxSize + 1]), from_lastname)) {
                k += 10;
                if (index == from_stop_blockindex and k + 1 >= util.u16ToUsize(from_offset)) return 4;
                continue;
            }
            if ((state & 0x01) != 1) return 2;
            @memcpy(b4[0..], from_block[k .. k + 4]);
            k += 4;
            from_file_start = util.b4ToU32(b4[0..]);
            @memcpy(b4[0..], from_block[k .. k + 4]);
            k += 4;
            from_file_stop = util.b4ToU32(b4[0..]);
            @memcpy(b2[0..], from_block[k .. k + 2]);
            from_file_offset = util.b2ToU16(b2[0..]);
            flag = true;
            break;
        }
        if (flag) break;
        @memcpy(b4[0..], from_block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) return 1;
        if (!block_mod.readblock(self, index, from_block[0..])) return 1;
    }

    var to_ba: [2]types.BlockArray = .{ .{}, .{} };
    var to_ba_used: usize = 0;
    var to_block_head: []u8 = undefined;
    var to_block_last: []u8 = undefined;
    var to_block_head_index: u32 = 0;
    var to_block_last_index: u32 = 0;

    var to_block = util.blockZero();
    if (!block_mod.readblock(self, to_blockindex, to_block[0..])) return 1;
    @memcpy(to_ba[0].block[0..], to_block[0..]);
    to_ba[0].blockindex = to_blockindex;
    to_ba[0].active = true;
    to_block_head = to_ba[0].block[0..];
    to_block_head_index = to_blockindex;
    to_ba_used = 1;

    @memcpy(b4[0..], to_block_head[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const to_stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], to_block_head[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    const to_offset = util.b2ToU16(b2[0..]);

    if (to_stop_blockindex == to_block_head_index) {
        to_block_last = to_block_head;
        to_block_last_index = to_block_head_index;
    } else {
        if (!block_mod.readblock(self, to_stop_blockindex, to_ba[1].block[0..])) return 1;
        to_ba[1].blockindex = to_stop_blockindex;
        to_ba[1].active = true;
        to_block_last = to_ba[1].block[0..];
        to_block_last_index = to_stop_blockindex;
        to_ba_used += 1;
    }

    flag = false;
    index = to_block_head_index;
    while (true) {
        var k: usize = types.BlockHead;
        var item: usize = 0;
        while (item < types.BlockItemMaxCount) : (item += 1) {
            k += 1;
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], to_block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..types.BlockNameMaxSize + 1]), to_lastname)) {
                k += 10;
                if (index == to_stop_blockindex and k + 1 >= util.u16ToUsize(to_offset)) {
                    flag = true;
                    break;
                }
                continue;
            }
            return 5;
        }
        if (flag) break;
        @memcpy(b4[0..], to_block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) return 1;
        if (!block_mod.readblock(self, index, to_block[0..])) return 1;
    }

    if (self.tmp.state == 0 and !txn.tmpstart(self, 1)) return 1;

    var block2 = util.blockZero();
    var blockindex2: u32 = 0;
    var new_to_offset: u16 = 0;
    if (util.u16ToUsize(to_offset) < types.BlockSize) {
        @memset(to_block_last[to_offset .. to_offset + 25], 0);
        to_block_last[to_offset] = 1;
        util.copyNameInto(to_block_last[to_offset + 1 .. to_offset + 1 + types.BlockNameMaxSize], to_lastname);
        new_to_offset = to_offset + 25;
        util.u16ToB2(new_to_offset, &b2);
        @memcpy(to_block_head[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
    } else {
        blockindex2 = block_mod.genblockindex(self);
        if (blockindex2 == 0) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        util.u32ToB4(to_block_last_index, &b4);
        @memcpy(block2[8..12], b4[0..]);
        block2[types.BlockHead] = 1;
        util.copyNameInto(block2[types.BlockHead + 1 .. types.BlockHead + 1 + types.BlockNameMaxSize], to_lastname);
        util.u32ToB4(blockindex2, &b4);
        @memcpy(to_block_last[4..8], b4[0..]);
        new_to_offset = types.BlockHead + 25;
        util.u16ToB2(new_to_offset, &b2);
        @memcpy(to_block_head[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
        @memcpy(to_block_head[types.BlockStopBlockindex .. types.BlockStopBlockindex + 4], b4[0..]);
    }

    var to_file_start: u32 = 0;
    var to_file_stop: u32 = 0;
    var to_file_offset: u16 = 0;

    if (from_file_start > 0) {
        to_file_offset = from_file_offset;
        var from_index = from_file_start;
        if (!block_mod.readblock(self, from_index, from_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        @memcpy(b4[0..], from_block[4..8]);
        var from_next_index = util.b4ToU32(b4[0..]);

        var new_blockindex = block_mod.genblockindex(self);
        if (new_blockindex == 0) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        to_file_start = new_blockindex;
        to_file_stop = new_blockindex;
        var prev_index: u32 = 0;
        var new_block = util.blockZero();
        while (true) {
            @memcpy(new_block[0..], from_block[0..]);
            util.u32ToB4(prev_index, &b4);
            @memcpy(new_block[8..12], b4[0..]);
            if (from_index == from_file_stop) {
                to_file_stop = new_blockindex;
                if (!block_mod.writeblock(self, new_blockindex, new_block[0..])) {
                    if (self.tmp.state == 1) txn.tmpstop(self);
                    return 1;
                }
                break;
            }
            prev_index = new_blockindex;
            new_blockindex = block_mod.genblockindex(self);
            util.u32ToB4(new_blockindex, &b4);
            @memcpy(new_block[4..8], b4[0..]);
            if (!block_mod.writeblock(self, prev_index, new_block[0..])) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 1;
            }
            from_index = from_next_index;
            if (!block_mod.readblock(self, from_index, from_block[0..])) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 1;
            }
            @memcpy(b4[0..], from_block[4..8]);
            from_next_index = util.b4ToU32(b4[0..]);
        }
    }

    if (util.u16ToUsize(to_offset) < types.BlockSize) {
        util.u32ToB4(to_file_start, &b4);
        @memcpy(to_block_last[to_offset + 15 .. to_offset + 19], b4[0..]);
        util.u32ToB4(to_file_stop, &b4);
        @memcpy(to_block_last[to_offset + 19 .. to_offset + 23], b4[0..]);
        util.u16ToB2(to_file_offset, &b2);
        @memcpy(to_block_last[to_offset + 23 .. to_offset + 25], b2[0..]);
    } else {
        util.u32ToB4(to_file_start, &b4);
        @memcpy(block2[types.BlockHead + 15 .. types.BlockHead + 19], b4[0..]);
        util.u32ToB4(to_file_stop, &b4);
        @memcpy(block2[types.BlockHead + 19 .. types.BlockHead + 23], b4[0..]);
        util.u16ToB2(to_file_offset, &b2);
        @memcpy(block2[types.BlockHead + 23 .. types.BlockHead + 25], b2[0..]);
        if (!block_mod.writeblock(self, blockindex2, block2[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
    }
    for (to_ba[0..to_ba_used]) |entry| {
        if (!entry.active) continue;
        if (!block_mod.writeblock(self, entry.blockindex, entry.block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
    }
    if (self.tmp.state == 1 and !txn.commit(self)) return 1;
    return 0;
}
